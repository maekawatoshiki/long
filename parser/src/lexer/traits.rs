use anyhow::Result;
use ast::token::{kind::TokenKind, Token};
use std::path::PathBuf;

pub trait LexerLike {
    /// Returns the next token.
    fn next(&mut self) -> Result<Option<Token>>;

    /// Peeks the next token.
    fn peek(&mut self) -> Result<Option<Token>>;

    /// Ungets the given `token`.
    fn unget(&mut self, token: Token);

    /// Skips the next token only if it matches the given kind.
    fn skip(&mut self, kind: TokenKind) -> bool;

    /// Returns the file path the lexer currently reads.
    fn filepath(&self) -> Option<&PathBuf>;
}

pub struct TokenStream {
    tokens: Vec<Token>,
    buf: Vec<Token>,
    cur: usize,
}

impl TokenStream {
    /// Creates a new `TokenStream`.
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens,
            buf: Vec::new(),
            cur: 0,
        }
    }
}

impl LexerLike for TokenStream {
    fn next(&mut self) -> Result<Option<Token>> {
        if let Some(tok) = self.buf.pop() {
            return Ok(Some(tok));
        }
        self.cur += 1;
        Ok(self.tokens.get(self.cur - 1).cloned())
    }

    fn peek(&mut self) -> Result<Option<Token>> {
        if let Some(tok) = self.buf.last().cloned() {
            return Ok(Some(tok));
        }
        Ok(self.tokens.get(self.cur).cloned())
    }

    fn unget(&mut self, token: Token) {
        self.buf.push(token);
    }

    fn skip(&mut self, kind: TokenKind) -> bool {
        if let Ok(Some(tok)) = self.next() {
            if tok.kind() == &kind {
                return true;
            }
            self.unget(tok);
        }
        false
    }

    fn filepath(&self) -> Option<&PathBuf> {
        None
    }
}

impl From<Vec<Token>> for TokenStream {
    fn from(tokens: Vec<Token>) -> Self {
        Self::new(tokens)
    }
}

#[test]
fn vec_tokens() {
    use ast::token::*;
    use sourceloc::SourceLoc;
    let tokens = vec![
        Token::new(
            kind::TokenKind::Ident("foo".to_string()),
            SourceLoc::new(0, 1),
        ),
        Token::new(
            kind::TokenKind::Int(kind::IntKind::Int(123)),
            SourceLoc::new(0, 5),
        ),
    ];
    let ll: &mut dyn LexerLike = &mut TokenStream::from(tokens);
    assert!(ll.next().unwrap().is_some());
    assert!(ll.next().unwrap().is_some());
    assert!(ll.next().unwrap().is_none());
}

#[test]
fn vec_tokens_skip() {
    use ast::token::*;
    use sourceloc::SourceLoc;
    let tokens = vec![
        Token::new(kind::SymbolKind::Add.into(), SourceLoc::new(0, 1)),
        Token::new(kind::IntKind::Int(123).into(), SourceLoc::new(0, 5)),
    ];
    let ll: &mut dyn LexerLike = &mut TokenStream::from(tokens);
    assert!(ll.skip(kind::SymbolKind::Add.into()));
    assert!(!ll.skip(kind::SymbolKind::Add.into()));
    assert!(ll.skip(kind::IntKind::Int(123).into()));
    assert!(!ll.skip(kind::IntKind::Int(123).into()));
}
