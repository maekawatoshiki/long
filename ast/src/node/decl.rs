use super::{
    expr::Expr,
    stmt::BlockStmt,
    ty::{FuncType, Type},
};

/// A declaration kind.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#1>
#[derive(Debug, Clone)]
pub enum Decl {
    FuncDef(FuncDef),
    SimpleDecl(SimpleDecl),
}

/// A function definition.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.fct.def.general#nt:function-definition>
#[derive(Debug, Clone)]
pub struct FuncDef {
    pub name: DeclaratorId,
    pub ty: FuncType,
    pub body: BlockStmt,
}

/// A simple declaration.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#nt:simple-declaration>
#[derive(Debug, Clone)]
pub struct SimpleDecl {
    pub names: Vec<DeclaratorId>,
    pub types: Vec<Type>,
    pub inits: Vec<Option<Expr>>,
}

/// A declarator id.
/// <https://timsong-cpp.github.io/cppwp/n3337/dcl.decl#nt:declarator-id>
#[derive(Debug, Clone)]
pub enum DeclaratorId {
    Ident(String),
}

impl From<FuncDef> for Decl {
    fn from(func_def: FuncDef) -> Self {
        Decl::FuncDef(func_def)
    }
}
