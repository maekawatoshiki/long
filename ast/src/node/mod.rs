use long_sourceloc::SourceLoc;

pub mod decl;
pub mod expr;
pub mod lit;
pub mod stmt;
pub mod ty;

/// Any kind of node with source location.
#[derive(Debug, Clone)]
pub struct Located<T> {
    /// Any kind of node.
    pub inner: T,

    /// The source location.
    pub loc: SourceLoc,
}

impl<T> Located<T> {
    /// Creates a new `Located`.
    pub fn new(inner: T, loc: SourceLoc) -> Self {
        Located { inner, loc }
    }

    /// Returns the inner node.
    pub fn inner(self) -> T {
        self.inner
    }

    /// Returns the reference to the inner node.
    pub fn inner_ref(&self) -> &T {
        &self.inner
    }

    /// Returns the source location.
    pub fn loc(&self) -> SourceLoc {
        self.loc
    }
}
