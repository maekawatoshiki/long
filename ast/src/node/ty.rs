#[derive(Debug, Clone, Copy)]
pub enum Type {
    Void,
    Bool,
    Char(Sign),
    Char16(Sign),
    Char32(Sign),
    WChar(Sign),
    Short(Sign),
    Int(Sign),
    Long(Sign),
    LongLong(Sign),
    Float,
    Double,
    // TODO: struct, union, enum, class...
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Sign {
    Signed,
    Unsigned,
}
