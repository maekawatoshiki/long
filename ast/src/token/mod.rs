pub mod kind;

use std::collections::HashSet;

use kind::TokenKind;
use long_sourceloc::SourceLoc;

#[derive(Debug, Clone)]
pub struct Token {
    /// The kind of the token.
    kind: TokenKind,

    /// `Some(...)` if any leading spaces occurs before the token.
    /// Line terminator characters are not leading spaces.
    leading_space: bool,

    /// The hideset of the token.
    // TODO: We should create `PPToken` (with a hideset) to simplify code.
    hideset: HashSet<String>,

    /// The source location of the token.
    loc: SourceLoc,
}

impl Token {
    /// Creates a new `Token`.
    #[inline]
    pub fn new(kind: TokenKind, loc: SourceLoc) -> Self {
        Self {
            kind,
            leading_space: false,
            hideset: HashSet::new(),
            loc,
        }
    }

    /// Sets the kind of the token and returns the token.
    #[inline]
    pub fn with_kind(mut self, kind: TokenKind) -> Self {
        self.kind = kind;
        self
    }

    /// Sets a leading space flag and returns the token.
    #[inline]
    pub fn set_leading_space(mut self, leading_space: bool) -> Self {
        self.leading_space = leading_space;
        self
    }

    /// Sets the hideset and returns the token.
    #[inline]
    pub fn with_hideset(mut self, hideset: HashSet<String>) -> Self {
        self.hideset = hideset;
        self
    }

    /// Modifies the hideset by `f` and returns the token.
    #[inline]
    pub fn with_hideset_modified<F>(mut self, f: F) -> Self
    where
        F: FnOnce(&mut HashSet<String>),
    {
        f(&mut self.hideset);
        self
    }

    /// Sets the source location and returns the token.
    #[inline]
    pub fn with_loc(mut self, loc: SourceLoc) -> Self {
        self.loc = loc;
        self
    }

    /// Returns the kind of the token.
    #[inline]
    pub fn kind(&self) -> &TokenKind {
        &self.kind
    }

    /// Returns the source location of the token.
    #[inline]
    pub fn loc(&self) -> &SourceLoc {
        &self.loc
    }

    /// Returns the leading space of the token.
    #[inline]
    pub fn leading_space(&self) -> bool {
        self.leading_space
    }

    /// Returns the hideset of the token.
    #[inline]
    pub fn hideset(&self) -> &HashSet<String> {
        &self.hideset
    }

    /// Converts the token to a string.
    #[inline]
    pub fn to_string(&self) -> String {
        self.kind.to_string()
    }
}

/// Converts the token stream to a string.
pub fn stringify(tokens: &[Token], ignore_leading_space: bool) -> String {
    tokens
        .iter()
        .fold("".to_string(), |acc, tok| {
            format!(
                "{}{}{}",
                acc,
                if !ignore_leading_space && tok.leading_space {
                    " "
                } else {
                    ""
                },
                tok.to_string()
            )
        })
        .trim_start() // Remove a leading space.
        .to_string()
}
