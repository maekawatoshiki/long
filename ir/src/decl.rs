use std::fmt;

use crate::{expr::Expr, name::Name, stmt::BlockStmt, ty::Type};
use id_arena::{Arena, Id};

#[derive(Debug)]
pub enum Decl<'a> {
    FuncDef(FuncDef<'a>),
    SimpleDecl(Vec<SimpleDecl<'a>>),
}

#[derive(Debug)]
pub struct FuncDef<'a> {
    pub name: &'a Name,
    pub sig: FuncSignature<'a>,
    pub body: BlockStmt<'a>,
    pub locals: Locals<'a>,
}

pub struct Locals<'a>(Arena<Local<'a>>);

#[derive(Debug)]
pub struct Local<'a> {
    pub name: &'a Name,
    pub ty: &'a Type<'a>,
}

#[derive(Debug)]
pub struct SimpleDecl<'a> {
    pub name: &'a Name,
    pub ty: &'a Type<'a>,
    pub init: Option<&'a Expr<'a>>,
}

/// A function signature (i.e. the return type and parameters of a function).
#[derive(Debug)]
pub struct FuncSignature<'a> {
    pub ret: &'a Type<'a>,
    pub params: Vec<Param<'a>>,
}

/// A parameter of a function.
#[derive(Debug)]
pub struct Param<'a> {
    pub name: &'a Name,
    pub ty: &'a Type<'a>,
}

impl<'a> Locals<'a> {
    pub fn new() -> Self {
        Locals(Arena::new())
    }

    pub fn alloc(&mut self, name: &'a Name, ty: &'a Type<'a>) -> Id<Local<'a>> {
        self.0.alloc(Local { name, ty })
    }

    pub fn iter(&self) -> impl Iterator<Item = (Id<Local<'a>>, &Local<'a>)> {
        self.0.iter()
    }
}

impl fmt::Debug for Locals<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Locals {{ ")?;
        for (i, local) in self.0.iter().enumerate() {
            if i > 0 {
                write!(f, ", ")?;
            }
            write!(f, "{:?}", local)?;
        }
        write!(f, " }}")
    }
}
