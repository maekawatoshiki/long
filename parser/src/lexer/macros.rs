use ast::token::Token;
use std::collections::HashMap;

/// Defined macros.
#[derive(Debug, Clone)]
pub struct Macros(HashMap<String, Macro>);

/// A macro body.
#[derive(Debug, Clone)]
pub enum Macro {
    /// Object-like macro.
    Obj(Vec<Token>),

    /// Function-like macro.
    Func(Vec<FuncMacroToken>),
}

/// A function-like macro token.
/// `Param(leading space, n)` represents a `n`th parameter.
/// `Vararg(leading space, n)` represents the `n`th parameter is `...`.
// TODO: Param and Vararg should not have leading space as bool.
#[derive(Debug, Clone)]
pub enum FuncMacroToken {
    Token(Token),
    Param(bool, usize),
    Vararg(bool, usize),
}

impl Macros {
    /// Creates a new `Macros`.
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    /// Adds a new object-like macro.
    pub fn add_obj_macro(&mut self, name: impl Into<String>, body: Vec<Token>) {
        self.0.insert(name.into(), Macro::Obj(body));
    }

    /// Adds a new function-like macro.
    pub fn add_func_macro(&mut self, name: impl Into<String>, body: Vec<FuncMacroToken>) {
        self.0.insert(name.into(), Macro::Func(body));
    }

    /// If `name` is a defined macro name, returns the corresponding macro body.
    pub fn find(&self, name: impl AsRef<str>) -> Option<&Macro> {
        self.0.get(name.as_ref())
    }

    pub fn remove(&mut self, name: impl AsRef<str>) {
        self.0.remove(name.as_ref());
    }
}
