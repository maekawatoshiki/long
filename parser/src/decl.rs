use crate::lexer::{traits::LexerLike, Error};
use crate::Parser;
use anyhow::Result;
use long_ast::node::decl::DeclaratorId;
use long_ast::node::ty::{Sign, Type};
use long_ast::node::{decl::Decl, Located};
use long_ast::token::kind::{KeywordKind, SymbolKind, TokenKind};
use long_ast::token::Token;

impl<'a, L: LexerLike> Parser<'a, L> {
    /// Parses a declaration sequence.
    /// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#1>
    pub(crate) fn parse_decl_seq(&mut self) -> Result<Located<Decl>> {
        // TODO
        self.parse_decl()
    }

    /// Parses a declaration.
    /// <https://timsong-cpp.github.io/cppwp/n3337/dcl.dcl#1>
    pub(crate) fn parse_decl(&mut self) -> Result<Located<Decl>> {
        // TODO
        self.parse_func_def()
    }

    /// Parses a function definition.
    /// <https://timsong-cpp.github.io/cppwp/n3337/dcl.fct.def.general#1>
    fn parse_func_def(&mut self) -> Result<Located<Decl>> {
        // function-definition:
        //     attribute-specifier-seq(opt) decl-specifier-seq(opt) declarator virt-specifier-seq(opt) function-body
        let Located {
            inner: decl_specifier,
            ..
        } = self.parse_decl_specifier()?;
        let mut declarator_id = None;
        let _declarator = self.parse_declarator(decl_specifier, &mut declarator_id);
        todo!()
    }

    /// Parses a declarator.
    /// <https://timsong-cpp.github.io/cppwp/n3337/dcl.decl#4>
    fn parse_declarator(
        &mut self,
        basety: Type,
        declarator_id: &mut Option<DeclaratorId>,
    ) -> Result<Type> {
        if self.lexer.skip(SymbolKind::Asterisk.into()) {
            // TODO: Parse attribute-specifier-seq(opt) cv-qualifier-seq(opt) here.
            return self.parse_declarator(Type::Pointer(Box::new(basety)), declarator_id);
        }

        if self.lexer.skip(SymbolKind::And.into()) {
            // TODO: Parse attribute-specifier-seq(opt) here.
            return self.parse_declarator(Type::LValueRef(Box::new(basety)), declarator_id);
        }

        if self.lexer.skip(SymbolKind::LAnd.into()) {
            // TODO: Parse attribute-specifier-seq(opt) here.
            return self.parse_declarator(Type::RValueRef(Box::new(basety)), declarator_id);
        }

        let tok = match self.lexer.next()? {
            Some(tok) => tok,
            None => return Ok(basety),
        };

        match tok.kind() {
            // TODO: Support more complex declarator-id.
            TokenKind::Ident(id) => *declarator_id = Some(DeclaratorId::Ident(id.clone())),
            _ => {
                self.lexer.unget(tok);
            }
        }

        Ok(basety)
    }

    /// Parses declaration specifiers.
    /// <https://timsong-cpp.github.io/cppwp/n3337/dcl.spec#nt:decl-specifier-seq>
    fn parse_decl_specifier(&mut self) -> Result<Located<Type>> {
        #[derive(PartialEq, Debug, Clone)]
        enum Size {
            Short,
            Long,
            LongLong,
            Unspecified,
        }

        #[derive(PartialEq, Debug, Clone)]
        enum PrimitiveType {
            Void,
            Bool,
            Char,
            Char16,
            Char32,
            WChar,
            Int,
            Float,
            Double,
        }

        let mut ty = None;
        let mut sign = None;
        let mut loc = None;
        let mut size = Size::Unspecified;

        fn err(tok: Token) -> Result<()> {
            Err(Error::Message(
                format!("Cannot specify '{}'", tok.kind().to_string()),
                *tok.loc(),
            )
            .into())
        }

        loop {
            let tok = match self.lexer.next()? {
                Some(tok) => tok,
                None => break,
            };

            if loc.is_none() {
                loc = Some(*tok.loc());
            }

            // TODO: Support more types: struct, union, enum, class, user defined types...

            match tok.kind() {
                TokenKind::Keyword(
                    KeywordKind::Void
                    | KeywordKind::Bool
                    | KeywordKind::Char
                    | KeywordKind::Char16T
                    | KeywordKind::Char32T
                    | KeywordKind::Int
                    | KeywordKind::Float
                    | KeywordKind::Double,
                ) if ty.is_some() => err(tok)?,
                TokenKind::Keyword(KeywordKind::Void) => ty = Some(PrimitiveType::Void),
                TokenKind::Keyword(KeywordKind::Bool) => ty = Some(PrimitiveType::Bool),
                TokenKind::Keyword(KeywordKind::Char) => ty = Some(PrimitiveType::Char),
                TokenKind::Keyword(KeywordKind::Char16T) => ty = Some(PrimitiveType::Char16),
                TokenKind::Keyword(KeywordKind::Char32T) => ty = Some(PrimitiveType::Char32),
                TokenKind::Keyword(KeywordKind::WcharT) => ty = Some(PrimitiveType::WChar),
                TokenKind::Keyword(KeywordKind::Int) => ty = Some(PrimitiveType::Int),
                TokenKind::Keyword(KeywordKind::Float) => ty = Some(PrimitiveType::Float),
                TokenKind::Keyword(KeywordKind::Double) => ty = Some(PrimitiveType::Double),
                TokenKind::Keyword(KeywordKind::Short) if size == Size::Unspecified => {
                    size = Size::Short
                }
                TokenKind::Keyword(KeywordKind::Short) => err(tok)?,
                TokenKind::Keyword(KeywordKind::Long) if size == Size::Unspecified => {
                    size = Size::Long
                }
                TokenKind::Keyword(KeywordKind::Long) if size == Size::Long => {
                    size = Size::LongLong;
                }
                TokenKind::Keyword(KeywordKind::Long) if size == Size::LongLong => {
                    err(tok)?;
                }
                TokenKind::Keyword(KeywordKind::Signed) if sign == Some(Sign::Signed) => {
                    // TODO: Had better raise a warning message.
                }
                TokenKind::Keyword(KeywordKind::Signed) if sign == Some(Sign::Unsigned) => {
                    err(tok)?
                }
                TokenKind::Keyword(KeywordKind::Unsigned) if sign == Some(Sign::Signed) => {
                    err(tok)?
                }
                TokenKind::Keyword(KeywordKind::Signed) => sign = Some(Sign::Signed),
                TokenKind::Keyword(KeywordKind::Unsigned) => sign = Some(Sign::Unsigned),
                _ => {
                    self.lexer.unget(tok);
                    break;
                }
            }
        }

        let loc = match loc {
            Some(loc) => loc,
            None => return Err(Error::UnexpectedEof.into()),
        };
        let sign = sign.unwrap_or(Sign::Signed);

        let ty = match ty {
            Some(PrimitiveType::Void) => Type::Void,
            Some(PrimitiveType::Bool) => Type::Bool,
            Some(PrimitiveType::Char) => Type::Char(sign),
            Some(PrimitiveType::Char16) => Type::Char16(sign),
            Some(PrimitiveType::Char32) => Type::Char32(sign),
            Some(PrimitiveType::WChar) => Type::WChar(sign),
            Some(PrimitiveType::Int) if size == Size::Short => Type::Short(sign),
            Some(PrimitiveType::Int) if size == Size::Long => Type::Long(sign),
            Some(PrimitiveType::Int) if size == Size::LongLong => Type::LongLong(sign),
            Some(PrimitiveType::Int) => Type::Int(sign),
            Some(PrimitiveType::Float) => Type::Float,
            Some(PrimitiveType::Double) => Type::Double,
            None => match size {
                Size::Short => Type::Short(sign),
                Size::Long => Type::Long(sign),
                Size::LongLong => Type::LongLong(sign),
                Size::Unspecified => Type::Int(sign),
            },
        };

        Ok(Located::new(ty, loc))
    }
}

macro_rules! decl_specifier_test {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            use crate::lexer::Lexer;
            let mut lexer = Lexer::new($src);
            let mut parser = Parser::new(&mut lexer);
            insta::assert_debug_snapshot!(parser.parse_decl_specifier());
        }
    };
}

decl_specifier_test!(decl_spec_empty, "");
decl_specifier_test!(decl_spec_bool, "bool");
decl_specifier_test!(decl_spec_char, "char");
decl_specifier_test!(decl_spec_char16, "char16_t");
decl_specifier_test!(decl_spec_char32, "char32_t");
decl_specifier_test!(decl_spec_unsigned_char, "unsigned char");
decl_specifier_test!(decl_spec_unsigned_char_2, "char unsigned");
decl_specifier_test!(decl_spec_wchar, "wchar_t");
decl_specifier_test!(decl_spec_short, "short");
decl_specifier_test!(decl_spec_int, "int");
decl_specifier_test!(decl_spec_long, "long");
decl_specifier_test!(decl_spec_long_long, "long long");
decl_specifier_test!(decl_spec_long_long_long, "long long long");
decl_specifier_test!(decl_spec_int_long, "int long");
decl_specifier_test!(decl_spec_long_int, "long int");
decl_specifier_test!(decl_spec_unsigned_long_int, "unsigned long int");
decl_specifier_test!(decl_spec_float, "float");
decl_specifier_test!(decl_spec_double, "double");

macro_rules! declarator_test {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            use crate::lexer::Lexer;
            let mut lexer = Lexer::new($src);
            let mut parser = Parser::new(&mut lexer);
            #[derive(Debug)]
            #[allow(dead_code)]
            struct Result {
                ty: Type,
                declarator_id: Option<DeclaratorId>,
            }
            let mut declarator_id = None;
            let ty = parser
                .parse_declarator(Type::Void, &mut declarator_id)
                .unwrap();
            insta::assert_debug_snapshot!(Result { ty, declarator_id })
        }
    };
}

declarator_test!(declarator_ptr, "*");
declarator_test!(declarator_many_ptr, "****");
declarator_test!(declarator_lvalref, "&");
declarator_test!(declarator_rvalref, "&&");
declarator_test!(declarator_id, "var");
