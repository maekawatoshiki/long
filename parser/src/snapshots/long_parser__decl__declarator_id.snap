---
source: parser/src/decl.rs
expression: "Result{ty, declarator_id,}"

---
Result {
    ty: Void,
    declarator_id: Some(
        Ident(
            "var",
        ),
    ),
}
